### PR TITLE
`TileOnGrid` support for `Tensor` input

### LINKED ISSUE #3379: Update TileOnGrid backend
**Is your feature request related to a problem? Please describe.**
Update `TileOnGrid` to support input/output type of `torch.Tensor` and `numpy` while its backend is `numpy` only.

### PR DESCRIPTION
**Pull request by drbeh**
Fixes #3379 

### Description
This PR updates `TileOnGrid` to support `torch.Tensor` through type conversion while the backend is `numpy`. It also updates the unittests to include `torch.Tenor` inputs on cpu and cuda device.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests`.

### CODE REVIEW

File: monai/apps/pathology/transforms/spatial/dictionary.py [hunk 1]
```diff
@@ -112,7 +110,9 @@ def __init__(
     def randomize(self, data: Any = None) -> None:
         self.seed = self.R.randint(10000)  # type: ignore
 
-    def __call__(self, data: Mapping[Hashable, np.ndarray]) -> Union[Dict[Hashable, np.ndarray], List[Dict]]:
+    def __call__(
```
Comment (Nic-Ma) [inline monai/apps/pathology/transforms/spatial/dictionary.py:113]: Please add the backend for dict transform also.

Thanks.

Comment (drbeh) [reply monai/apps/pathology/transforms/spatial/dictionary.py:115]: Thanks! Done!

**Review by Nic-Ma — APPROVED** [review]

File: monai/apps/pathology/transforms/spatial/array.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
 
 from monai.config.type_definitions import NdarrayOrTensor
 from monai.transforms.transform import Randomizable, Transform
+from monai.utils import convert_data_type, convert_to_dst_type
 from monai.utils.enums import TransformBackends
 
 __all__ = ["SplitOnGrid", "TileOnGrid"]
```
Comment (wyli) [inline monai/apps/pathology/transforms/spatial/array.py:23]: looks like `SplitOnGrid(2, 2)(torch.arange(12).reshape(1, 3, 4))` works fine but `SplitOnGrid(2, 2)(np.arange(12).reshape(1, 3, 4))` doesn't work, could you help fix it here? #3378 was merged too quickly

File: monai/apps/pathology/transforms/spatial/array.py [hunk 3]
```diff
@@ -185,37 +188,39 @@ def randomize(self, img_size: Sequence[int]) -> None:
...
                 [[0, 0], [pad_h // 2, pad_h - pad_h // 2], [pad_w // 2, pad_w - pad_w // 2]],
                 constant_values=self.background_val,
             )
 
         # extact tiles
-        x_step, y_step = self.step, self.step
-        x_size, y_size = self.tile_size, self.tile_size
-        c_len, x_len, y_len = image.shape
-        c_stride, x_stride, y_stride = image.strides
+        h_step, w_step = self.step, self.step
```
Comment (wyli) [inline monai/apps/pathology/transforms/spatial/array.py:213]: I feel the previous variable `x_step` is better than `h_step`, `x` indicates the axis direction, `h` indicates `height` which is the length of `x` direction

File: monai/apps/pathology/transforms/spatial/array.py (hunk 2 above)
Comment (drbeh) [reply monai/apps/pathology/transforms/spatial/array.py:23]: @wyli, sorry I missed this. I will fix it now.

File: monai/apps/pathology/transforms/spatial/array.py (hunk 3 above)
Comment (drbeh) [reply monai/apps/pathology/transforms/spatial/array.py:213]: Sure, I can fix this too. sorry that I didn't see your comments before merging.

File: monai/apps/pathology/transforms/spatial/array.py (hunk 2 above)
Comment (drbeh) [reply monai/apps/pathology/transforms/spatial/array.py:23]: https://github.com/Project-MONAI/MONAI/pull/3386

File: monai/apps/pathology/transforms/spatial/array.py (hunk 3 above)
Comment (drbeh) [reply monai/apps/pathology/transforms/spatial/array.py:213]: https://github.com/Project-MONAI/MONAI/pull/3386

**Pull request merged**